service/registration.go

```diff
@@ -12,77 +17,283 @@ type RegistrationWrapper interface {
 	WaitRegistration()                 // WaitRegistration wait for the first registration
 	Credentials() (Credentials, error) // Credentials return the current credentials or an error if the service is not registred yet
 	UUID() string                      // UUID return the host UUID
+	Stop()                             // Stop the registration and free the resources (stop goroutines/close channels)
 }
 
 // Registration is the RegistrationWrapper implementation used by the Register method
 type Registration struct {
-	credChan       chan (Credentials)
-	readyChan      chan bool
-	ready          bool
-	uuid           string
-	mutex          sync.Mutex
-	curCredentials *Credentials
+	// Lifecycle management
+	wg          *sync.WaitGroup
+	cancel      context.CancelFunc
+	shutdownErr error
+
+	credsWatcher       chan Credentials
```

Comment:
Maybe a few comments to explain the purpose of these channels?

Reply:
Yep